src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncIteratorMethodToStateMachineRewriter.cs

```diff
@@ -292,70 +297,71 @@ private BoundExpressionStatement SetDisposeMode(bool value)
         ///
         /// Some `finally` clauses may have already been rewritten and extracted to a plain block (<see cref="AsyncExceptionHandlerRewriter"/>).
         /// In those cases, we saved the finally-entry label in <see cref="BoundTryStatement.FinallyLabelOpt"/>.
         /// </summary>
         public override BoundNode VisitTryStatement(BoundTryStatement node)
         {
-            _previousDisposalLabels ??= new ArrayBuilder<LabelSymbol>();
-            _previousDisposalLabels.Push(_enclosingFinallyEntryOrFinallyExitOrExitLabel);
-
-            var finallyExit = F.GenerateLabel("finallyExit");
-            _previousDisposalLabels.Push(finallyExit);
-
-            if (node.FinallyBlockOpt != null)
+            var savedDisposalLabel = _currentDisposalLabel;
+            bool changedDisposalLabel;
+            if (node.FinallyBlockOpt is object)
             {
                 var finallyEntry = F.GenerateLabel("finallyEntry");
-                _enclosingFinallyEntryOrFinallyExitOrExitLabel = finallyEntry;
+                _currentDisposalLabel = finallyEntry;
+                changedDisposalLabel = true;
 
                 // Add finallyEntry label:
                 //  try
                 //  {
                 //      ...
                 //      finallyEntry:
-
-                // Add finallyExit label:
-                //  finally
-                //  {
-                //      ...
-                //      finallyExit:
                 //  }
+
                 node = node.Update(
                     tryBlock: F.Block(node.TryBlock, F.Label(finallyEntry)),
-                    node.CatchBlocks, F.Block(node.FinallyBlockOpt, F.Label(finallyExit)), node.FinallyLabelOpt, node.PreferFaultHandler);
+                    node.CatchBlocks, node.FinallyBlockOpt, node.FinallyLabelOpt, node.PreferFaultHandler);
+            }
+            else if (node.FinallyLabelOpt is object)
+            {
+                _currentDisposalLabel = node.FinallyLabelOpt;
+                changedDisposalLabel = true;
             }
-            else if ((object)node.FinallyLabelOpt != null)
+            else
             {
-                _enclosingFinallyEntryOrFinallyExitOrExitLabel = node.FinallyLabelOpt;
+                changedDisposalLabel = false;
             }
 
             var result = (BoundStatement)base.VisitTryStatement(node);
 
-            // While inside the try and catch blocks, we'll use the current finallyEntry label for disposal
-            // As soon as we close the try and catch blocks (ie. possibly enter the finally block), we'll use the finallyExit label for disposal (restored/popped from the stack by CloseTryCatchBlocks)
-            Debug.Assert(_enclosingFinallyEntryOrFinallyExitOrExitLabel == finallyExit);
-
-            // When exiting the try statement, we restore the previous disposal label.
-            _enclosingFinallyEntryOrFinallyExitOrExitLabel = _previousDisposalLabels.Pop();
+            if (changedDisposalLabel)
```

Comment:
>changedDisposalLabel [](start = 16, length = 20)

You don't need to track whether or not it changed. You can just unconditionally set it to what it used to be.

Reply:
Will do. Thanks